### DISCUSSION TITLE
Discussion: postfix relational patterns

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#4110] [Language Ideas]
In C# 9.0 range checks can be written using multiple relational patterns:
```cs
n is >= 10 and <= 20
```
But that might be difficult to reason about at the first glance duo to operators being directed in the same direction. 

The idea here is to allow postfix relational patterns so that following is permitted:
```cs
n is 10 <= and <= 20
```
I think this reads better as we have both range bounds in either side of the whole pattern.

Since we allow `and` and `or` as identifiers, the following valid code would be ambigious with this change:
```cs
n is a < and > b
```
as it could be interpreted as either a declaration pattern or a range check; though I think it is an acceptable breaking change.

This would be an alternative to the proposal at https://github.com/dotnet/csharplang/issues/4108 which is enabling something similar.

cc @CyrusNajmabadi

**Comment by yaakov-h** [comment]
I really don't like how the first part (`>= 10`) gets inverted (`10 <=`) and found it quite hard to parse mentally on the first few passes. Are there other use-cases where this is definitively clearer than alternatives?

**Reply by 333fred** [reply]
I would agree: I don't think this version reads particularly well, and adds a bunch of confusion for (imo) net negative readability.

**Reply by alrz** [reply]
> I really don't like how the first part gets inverted

In comparison with https://github.com/dotnet/csharplang/issues/4108, the current syntax is actually look inverted and that's what makes it hard to read :)

**Reply by 333fred** [reply]
Well, that's a different syntax altogether, and reads differently. The existing pattern syntaxes read quite well to me:
`n is >= 10 and <= 20` reads as "o is greater than or equal to 10 and less than or equal to 20". This proposal would is `n is 10 <= and <= 20`, which would read as "o is 10 less than or equal to and greater than or equal to 20". I don't know what that means, but I find it confusing. The other proposal is `10 <= o <= 20`, which would read "10 is less than or equal o which is less than or equal to 20". It reads fine because of the positioning of the conditions: the same does not apply to this proposal.

**Reply by alrz** [reply]
It's not supposed to improve how it literally "reads", we're using operators in opposite directions and that I think is what makes it more clear when you're writing range checks.

I agree it is pretty subjective whether you prefer postfix or prefix syntax, but my opinion is that if we're going to consider #4108, a pattern form would better fit the use case now that we already have a feature in place which is supposed to simplify this exact code pattern.

**Reply by saint4eva** [reply]
> I agree it is pretty subjective whether you prefer postfix or prefix syntax, but my opinion is that if we're going to consider #4108, a pattern form would better fit the use case now that we already have a feature in place which is supposed to simplify this exact code pattern.

Not only that the proposal is very subjective, it is confusing also. I do not think that it adds any value.

**Comment by CyrusNajmabadi** [comment]
> This would be an alternative to the proposal at #4108 which is enabling something similar.

Sort of :)  However, #4108 is really about expressions (not patterns), where i think there's a ton of value.

**Reply by alrz** [reply]
The reason I mentioned it is that patterns are meant to solve these kind of problems. I don't think adding a separate syntax for the exact same thing is a good Idea.

>where i think there's a ton of value.

What other scenarios would it solve except matching with non-constants? And for that I think there's room for improvement in patterns themselves.
